larq/optimizers_v2.py

```diff
@@ -5,6 +5,84 @@
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class OptimizerGroup(tf.keras.optimizers.Optimizer):
+    """Group of a full-precision and binary optimizer.
+
+    #TODO: Do we want to abstract this to `n` optimizers, not specifically one fp and
+    one bin? (I personally prefer the user-side simplicity of this current approach.)
+    """
+
+    def __init__(self, bin_optimizer, fp_optimizer, name="Group"):
+        super().__init__(name=name)  # TODO: Do we need to pass **kwargs?
+
+        type_err_msg = "Expected tf.keras.optimizers.Optimizer for `{}`, received `{}.`"
+        if not isinstance(bin_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(type_err_msg.format("bin_optimizer", type(bin_optimizer)))
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(type_err_msg.format("fp_optimizer", type(bin_optimizer)))
+
+        self.bin_optimizer = bin_optimizer
+        self.fp_optimizer = fp_optimizer
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        bin_grads_and_vars, fp_grads_and_vars = [], []
+
+        for grad, var in grads_and_vars:
+            if self.is_binary(var):
```

Comment:
I would like the decision rule to be modular, possibly even an argument given to `__init__()`, instead of hardcoded here.

Reply:
(See discussion [here](https://github.com/larq/larq/pull/304/files/4ce2e10b4c5ecbffc5f9ec1353f7cf5ee0a625d7#r341977897).)